src/core/service.c

```diff
@@ -2441,6 +2490,109 @@ static int service_deserialize_item(Unit *u, const char *key, const char *value,
+                                state = STATE_EXEC_COMMAND_PATH;
+                                break;
+                        case STATE_EXEC_COMMAND_PATH:
+                                path = temp;
+                                temp = NULL;
+                                state = STATE_EXEC_COMMAND_ARGS;
+
+                                assert(path_is_absolute(path));
+                                break;
+                        case STATE_EXEC_COMMAND_ARGS:
+                                r = strv_extend(&argv, temp);
+                                if (r < 0)
+                                        return log_oom();
+                                break;
+                        default:
+                                assert_not_reached("Unknown error at deserialization of exec command");
+                                break;
+                        }
+                }
+
+                args = strv_join(argv, " ");
+                if (!args)
+                        return log_oom();
+
+                /* Let's check whether exec command on given offset matches data that we just deserialized */
+                for (command = s->exec_command[id], i = 0; command; command = command->command_next, i++) {
+                        _cleanup_free_ char *candidate_args = NULL;
+
+                        if (i != index)
+                                continue;
+
+                        candidate_args = strv_join(command->argv, " ");
+                        if (!candidate_args)
+                                return log_oom();
+
+                        found = streq(args, candidate_args) && streq(command->path, path);
+                        break;
+                }
+
+                if (!found) {
+                        /* Command at the index we serialized is different, let's look for command that exactly
+                         * matches but is on different index. If there is no such command we will resume execution
+                         * from the serialized index.
+                         */
+                        for (command = s->exec_command[id]; command; command = command->command_next) {
+                                _cleanup_free_ char *candidate_args = NULL;
+
+                                candidate_args = strv_join(command->argv, " ");
+                                if (!candidate_args)
+                                        return log_oom();
+
+                                if (streq(args, candidate_args) && streq(command->path, path))
+                                        break;
+                        }
+                }
+
+                if (command && control)
+                        s->control_command = command;
+                else if (command)
+                        s->main_command = command;
+                else
+                        log_unit_notice(u, "Previously executed command vanished from the unit file, execution of the command list won't be resumed.");
```

Comment:
given how complex this is I think this bit should be mostly moved to its own function.